Code/Framework/AzCore/Tests/EBus/EBusSharedDispatchMutexTests.cpp

```diff
@@ -0,0 +1,293 @@
+        AZStd::semaphore m_disconnectSemaphore;
+
+        AZStd::atomic_int m_numDisconnects = 0;
+
+        SharedDispatchRequestHandler()
+        {
+            // Reinitialize these for every test.
+            m_totalRecursiveQueriesInProgress = 0;
+            m_totalRecursiveQueriesCompleted = 0;
+        }
+
+        ~SharedDispatchRequestHandler() override
+        {
+            SharedDispatchRequestBus::Handler::BusDisconnect();
+        }
+
+        void Connect()
+        {
+            SharedDispatchRequestBus::Handler::BusConnect();
+        }
+
+        void Disconnect()
+        {
+            // Signal that the thread is running and has at least made it this far.
+            m_disconnectSemaphore.release();
+
+            SharedDispatchRequestBus::Handler::BusDisconnect();
+            m_numDisconnects++;
+        }
+
+        void RecursiveQuery(int32_t numRecursions = 5) override
+        {
+            if (numRecursions <= 0)
+            {
+                // At the end of the recursion, signal the syncSemaphore that we've reached the end of the recursion.
+                // We'll use this as a way to guarantee that all our threads have reached this point at the same time.
+                m_syncSemaphore.release();
+
+                // Block on the querySemaphore. This won't get released until every thread has released the syncSemaphore.
+                m_querySemaphore.acquire();
+
+                // Track that we've completed the query successfully.
+                m_totalRecursiveQueriesCompleted++;
+                return;
+            }
+
+            // Recursively call the EBus a fixed number of times, and keep track of how many times we've successfully recursed.
+            m_totalRecursiveQueriesInProgress++;
+            SharedDispatchRequestBus::Broadcast(&SharedDispatchRequestBus::Events::RecursiveQuery, numRecursions - 1);
+            m_totalRecursiveQueriesInProgress--;
+        }
+    };
+
+    class EBusSharedDispatchMutexTestFixture
+        : public AllocatorsFixture
+    {
+    public:
+
+        EBusSharedDispatchMutexTestFixture()
+        {
+            SharedDispatchRequestBus::GetOrCreateContext();
+        }
+    };
+
+    TEST_F(EBusSharedDispatchMutexTestFixture, RecursiveBusCallsOnSingleThreadWorks)
+    {
+        // Verify that multiple nested bus calls to the same bus on the same thread works without deadlocks.
+
+        constexpr int32_t TotalRecursiveQueries = 10;
+        SharedDispatchRequestHandler handler;
+        handler.Connect();
+
+        // This is a single-threaded test, so we don't need the recursive query to block before returning.
+        handler.m_querySemaphore.release();
+
+        SharedDispatchRequestBus::Broadcast(&SharedDispatchRequestBus::Events::RecursiveQuery, TotalRecursiveQueries);
+        EXPECT_EQ(handler.m_totalRecursiveQueriesInProgress, 0);
+        EXPECT_EQ(handler.m_totalRecursiveQueriesCompleted, 1);
+        handler.Disconnect();
+    }
```

Comment:
The handler.syncSemaphore isn't acquired in this test, so handler will have a value of -1 at the end of the test for that semaphore.

Reply:
Yeah, the syncSemaphore just isn't needed in this test. I could add an acquire() at the end just to balance it out if you'd like.

Reply:
Added